### PR TITLE
impl of cmd create test report

### PR DESCRIPTION
**Pull request by rioliu-rh**
add first version of cmd `create-test-report`, including some core modules impls and unit test cases
- cli cmd, group command and sub command
- worksheet manager to duplicate, update test report
- advisory manager to get jira issues (flaw bugs are on bugzilla)
- jira manager to get/update/create jira issues, util funcs to check issue is critical/cve/blocker
- logging init, disable debug logging for some internal loggers.
- declared some exceptions, raise from app level, can be handled by higher level app
- declared constants for frequently used str

- cli running result:
```
# help
cli --help                          
Usage: cli [OPTIONS] COMMAND [ARGS]...

Options:
  -r, --release TEXT  z-stream releaes version  [required]
  -v, --debug         enable debug logging
  --help              Show this message and exit.

Commands:
  create-test-report  Create test report for z-stream release
```
```
# required option is missed
cli create-test-report
Usage: cli [OPTIONS] COMMAND [ARGS]...
Try 'cli --help' for help.

Error: Missing option '-r' / '--release'.
```
```
# w/o debug flag
cli -r 4.12.16 create-test-report 
2023-05-05T16:38:07Z: INFO: local config is loaded
2023-05-05T16:41:47Z: INFO: ocp build data of 4.12.16 is downloaded
2023-05-05T16:41:51Z: INFO: build info is updated
2023-05-05T16:41:52Z: INFO: advisory info is updated
2023-05-05T16:41:53Z: INFO: jira info is updated
2023-05-05T16:43:21Z: INFO: bugs to be verified are updated
2023-05-05T16:43:21Z: INFO: new test report is created: https://docs.google.com/spreadsheets/d/1Xv8qfYUwp61lOOMQaXNO7wnlo9b8mXACgug49cZHCTM#gid=1040444075
```
```
# w/ debug flag
cli -r 4.12.16 -v create-test-report
2023-05-05T16:51:15Z: INFO: local config is loaded
2023-05-05T16:51:16Z: INFO: ocp build data of 4.12.16 is downloaded
2023-05-05T16:51:19Z: INFO: build info is updated
2023-05-05T16:51:19Z: DEBUG: build info:
aarch64: 4.12.0-0.nightly-arm64-2023-05-03-200750
ppc64le: 4.12.0-0.nightly-ppc64le-2023-05-03-200749
s390x: 4.12.0-0.nightly-s390x-2023-05-03-200748
x86_64: 4.12.0-0.nightly-2023-05-04-010834

2023-05-05T16:51:20Z: INFO: advisory info is updated
2023-05-05T16:51:20Z: DEBUG: advisory info:
extras: 113789
image: 113788
metadata: 113790
rpm: 113787
microshift: 113791

2023-05-05T16:51:20Z: INFO: jira info is updated
2023-05-05T16:51:20Z: DEBUG: jira info:
ART-6731
2023-05-05T16:52:32Z: DEBUG: updating jira issue OCPBUGS-6572 ...
2023-05-05T16:52:32Z: DEBUG: jira issue OCPBUGS-6572 status is Verified, skipping
2023-05-05T16:52:33Z: DEBUG: updating jira issue OCPBUGS-11559 ...
2023-05-05T16:52:33Z: DEBUG: jira issue OCPBUGS-11559 is ON_QA, updating
2023-05-05T16:52:34Z: DEBUG: updating jira issue OCPBUGS-11844 ...
2023-05-05T16:52:34Z: DEBUG: jira issue OCPBUGS-11844 is ON_QA, updating
2023-05-05T16:52:35Z: DEBUG: updating jira issue OCPBUGS-11972 ...
2023-05-05T16:52:35Z: DEBUG: jira issue OCPBUGS-11972 status is Verified, skipping
2023-05-05T16:52:36Z: DEBUG: updating jira issue OCPBUGS-11993 ...
2023-05-05T16:52:36Z: DEBUG: jira issue OCPBUGS-11993 status is Verified, skipping
2023-05-05T16:52:36Z: DEBUG: updating jira issue OCPBUGS-12159 ...
2023-05-05T16:52:36Z: DEBUG: jira issue OCPBUGS-12159 status is Verified, skipping
2023-05-05T16:52:37Z: DEBUG: updating jira issue OCPBUGS-12199 ...
2023-05-05T16:52:37Z: DEBUG: jira issue OCPBUGS-12199 status is Verified, skipping
2023-05-05T16:52:38Z: DEBUG: updating jira issue OCPBUGS-12265 ...
2023-05-05T16:52:38Z: DEBUG: jira issue OCPBUGS-12265 status is Verified, skipping
2023-05-05T16:52:39Z: DEBUG: updating jira issue OCPBUGS-12361 ...
2023-05-05T16:52:39Z: DEBUG: jira issue OCPBUGS-12361 is ON_QA, updating
2023-05-05T16:52:40Z: DEBUG: updating jira issue OCPBUGS-12440 ...
2023-05-05T16:52:40Z: DEBUG: jira issue OCPBUGS-12440 status is Verified, skipping
2023-05-05T16:52:41Z: DEBUG: updating jira issue OCPBUGS-12473 ...
2023-05-05T16:52:41Z: DEBUG: jira issue OCPBUGS-12473 status is Verified, skipping
2023-05-05T16:52:42Z: DEBUG: updating jira issue OCPBUGS-12476 ...
2023-05-05T16:52:42Z: DEBUG: jira issue OCPBUGS-12476 status is Verified, skipping
2023-05-05T16:52:42Z: DEBUG: updating jira issue OCPBUGS-12477 ...
2023-05-05T16:52:42Z: DEBUG: jira issue OCPBUGS-12477 status is Verified, skipping
2023-05-05T16:52:43Z: DEBUG: updating jira issue OCPBUGS-12688 ...
2023-05-05T16:52:43Z: DEBUG: jira issue OCPBUGS-12688 is ON_QA, updating
2023-05-05T16:52:44Z: DEBUG: updating jira issue OCPBUGS-12768 ...
2023-05-05T16:52:44Z: DEBUG: jira issue OCPBUGS-12768 status is Verified, skipping
2023-05-05T16:52:45Z: DEBUG: updating jira issue OCPBUGS-12993 ...
2023-05-05T16:52:45Z: DEBUG: jira issue OCPBUGS-12993 is ON_QA, updating
2023-05-05T16:52:46Z: DEBUG: updating jira issue OCPBUGS-1753 ...
2023-05-05T16:52:46Z: DEBUG: jira issue OCPBUGS-1753 status is Verified, skipping
2023-05-05T16:52:47Z: DEBUG: updating jira issue OCPBUGS-4276 ...
2023-05-05T16:52:47Z: DEBUG: jira issue OCPBUGS-4276 status is Verified, skipping
2023-05-05T16:52:48Z: DEBUG: updating jira issue OCPBUGS-6888 ...
2023-05-05T16:52:48Z: DEBUG: jira issue OCPBUGS-6888 status is Verified, skipping
2023-05-05T16:52:48Z: DEBUG: updating jira issue OCPBUGS-7400 ...
2023-05-05T16:52:48Z: DEBUG: jira issue OCPBUGS-7400 is ON_QA, updating
2023-05-05T16:52:49Z: DEBUG: updating jira issue OCPBUGS-13087 ...
2023-05-05T16:52:49Z: DEBUG: jira issue OCPBUGS-13087 status is Verified, skipping
2023-05-05T16:52:50Z: DEBUG: updating jira issue OCPBUGS-13088 ...
2023-05-05T16:52:50Z: DEBUG: jira issue OCPBUGS-13088 status is Verified, skipping
2023-05-05T16:52:51Z: DEBUG: updating jira issue OCPBUGS-13089 ...
2023-05-05T16:52:51Z: DEBUG: jira issue OCPBUGS-13089 status is Verified, skipping
2023-05-05T16:52:51Z: INFO: bugs to be verified are updated
2023-05-05T16:52:51Z: INFO: new test report is created: https://docs.google.com/spreadsheets/d/1Xv8qfYUwp61lOOMQaXNO7wnlo9b8mXACgug49cZHCTM#gid=1040444075
```
- unittest result
```
Running tests (unittest): /Users/rio.liu/coderepo/release-tests/tests/test_advisory_mgr.py
Running tests: /Users/rio.liu/coderepo/release-tests/tests/test_advisory_mgr.py::TestAdvisoryManager::test_get_jira_issues
/Users/rio.liu/coderepo/release-tests/tests/test_advisory_mgr.py::TestAdvisoryManager::test_init
./test_advisory_mgr.py::TestAdvisoryManager::test_get_jira_issues Passed

./test_advisory_mgr.py::TestAdvisoryManager::test_init Passed

Running tests (unittest): /Users/rio.liu/coderepo/release-tests/tests/test_config_store.py
Running tests: /Users/rio.liu/coderepo/release-tests/tests/test_config_store.py::TestConfigStore::test_get_advisories
/Users/rio.liu/coderepo/release-tests/tests/test_config_store.py::TestConfigStore::test_get_cadidate_builds
/Users/rio.liu/coderepo/release-tests/tests/test_config_store.py::TestConfigStore::test_get_email_contact
/Users/rio.liu/coderepo/release-tests/tests/test_config_store.py::TestConfigStore::test_get_jira
/Users/rio.liu/coderepo/release-tests/tests/test_config_store.py::TestConfigStore::test_get_owner
/Users/rio.liu/coderepo/release-tests/tests/test_config_store.py::TestConfigStore::test_get_report_template
/Users/rio.liu/coderepo/release-tests/tests/test_config_store.py::TestConfigStore::test_get_slack_contact
/Users/rio.liu/coderepo/release-tests/tests/test_config_store.py::TestConfigStore::test_init
./test_config_store.py::TestConfigStore::test_get_advisories Passed

./test_config_store.py::TestConfigStore::test_get_cadidate_builds Passed

./test_config_store.py::TestConfigStore::test_get_email_contact Passed

./test_config_store.py::TestConfigStore::test_get_jira Passed

./test_config_store.py::TestConfigStore::test_get_owner Passed

./test_config_store.py::TestConfigStore::test_get_report_template Passed

./test_config_store.py::TestConfigStore::test_get_slack_contact Passed

./test_config_store.py::TestConfigStore::test_init Passed

Running tests (unittest): /Users/rio.liu/coderepo/release-tests/tests/test_jira_mgr.py::TestJiraManager
Running tests: /Users/rio.liu/coderepo/release-tests/tests/test_jira_mgr.py::TestJiraManager::test_create_issue
/Users/rio.liu/coderepo/release-tests/tests/test_jira_mgr.py::TestJiraManager::test_get_issue_fields
/Users/rio.liu/coderepo/release-tests/tests/test_jira_mgr.py::TestJiraManager::test_init
/Users/rio.liu/coderepo/release-tests/tests/test_jira_mgr.py::TestJiraManager::test_invalid_token
/Users/rio.liu/coderepo/release-tests/tests/test_jira_mgr.py::TestJiraManager::test_update_issue
./test_jira_mgr.py::TestJiraManager::test_create_issue Skipped: null


./test_jira_mgr.py::TestJiraManager::test_get_issue_fields Passed

./test_jira_mgr.py::TestJiraManager::test_init Skipped: null


./test_jira_mgr.py::TestJiraManager::test_invalid_token Skipped: null


./test_jira_mgr.py::TestJiraManager::test_update_issue Passed

Running tests (unittest): /Users/rio.liu/coderepo/release-tests/tests/test_worksheet_mgr.py::TestWorksheetManager
Running tests: /Users/rio.liu/coderepo/release-tests/tests/test_worksheet_mgr.py::TestWorksheetManager::test_init
/Users/rio.liu/coderepo/release-tests/tests/test_worksheet_mgr.py::TestWorksheetManager::test_report
./test_worksheet_mgr.py::TestWorksheetManager::test_init Passed

./test_worksheet_mgr.py::TestWorksheetManager::test_report Passed


```

### CODE REVIEW

File: oar/core/worksheet_mgr.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+            try:
+                template = self._doc.worksheet("template")
+            except WorksheetNotFound as we:
+                raise WorksheetException("cannot find template worksheet") from we
+
+            # check report worksheet exists or not, if yes, skip duplicating
+            try:
+                existing_sheet = self._doc.worksheet(self._cs.release)
+                self._report = TestReport(existing_sheet, self._cs)
+            except WorksheetNotFound:
```
Comment (kuiwang02) [inline oar/core/worksheet_mgr.py:67]: @rioliu-rh how about if it is not WorksheetNotFound? we should raise exception.

File: oar/core/worksheet_mgr.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+            # update jira info
+            self._report.update_jira_info(self._cs.get_jira_ticket())
+            logger.info("jira info is updated")
+            logger.debug(f"jira info:\n{self._cs.get_jira_ticket()}")
+
+            # update on_qa bugs list
+            am = AdvisoryManager(self._cs)
+            self._report.generate_bug_list(am.get_jira_issues())
+
+        except GSpreadException as ge:
```
Comment (kuiwang02) [inline oar/core/worksheet_mgr.py:98]: @rioliu-rh same above. how about other exception?

File: oar/core/worksheet_mgr.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+                }
+            ],
+            value_input_option=gspread.utils.ValueInputOption.user_entered,
+        )
+        # TODO: highlight the cell if the issue is critical
+        # TODO: send email/slack notification
+
+        logger.info("bugs to be verified are updated")
+
+    def update_bug_list(self, jira_issues: []):
```
Comment (kuiwang02) [inline oar/core/worksheet_mgr.py:298]: @rioliu-rh will it be done? reminder: incremental update.

File: oar/cli/cmd_create_test_report.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    """
+    Create test report for z-stream release
+    """
+    # get config store from context
+    cs = ctx.obj["cs"]
+    # init worksheet manager
+    try:
+        wm = WorksheetManager(cs)
+        report = wm.create_test_report()
+    except WorksheetException as we:
```
Comment (kuiwang02) [inline oar/cli/cmd_create_test_report.py:21]: @rioliu-rh how about other exception?

File: oar/core/worksheet_mgr.py (hunk 1 above)
Comment (rioliu-rh) [reply oar/core/worksheet_mgr.py:67]: the logic is check target worksheet exists or not, if yes, don't duplicate from template, just update it with build data

File: oar/core/worksheet_mgr.py (hunk 3 above)
Comment (rioliu-rh) [reply oar/core/worksheet_mgr.py:298]: yes, this func will update bug list with existing cell data

File: oar/core/worksheet_mgr.py (hunk 1 above)
Comment (kuiwang02) [reply oar/core/worksheet_mgr.py:67]: if other exception is raised, the code does not catch it.

File: oar/cli/cmd_create_test_report.py (hunk 4 above)
Comment (rioliu-rh) [reply oar/cli/cmd_create_test_report.py:21]: catch all the exception in `create_test_report` raise as `WorksheetException`

**Comment by rioliu-rh** [comment]
@kuiwang02 checked-in the fixes of error handling logic

**Comment by rioliu-rh** [comment]
/retitle OCPQE-17513 impl of cmd create test report

**Pull request merged**